### PR TITLE
Element search by 'class chain' locator type 

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
The idea for this PR is borrowed from Adobe Flex open-source testing framework, where xpath queries are not supported natively. I did it, because I observe many people still build xpath queries, like this

```
/XCUIElementTypeApplication/XCUIElementTypeWindow/XCUIElementTypeOther/XCUIElementTypeOther/XCUIElementTypeOther[2]/XCUIElementTypeOther/XCUIElementTypeOther/XCUIElementTypeOther/XCUIElementTypeOther/XCUIElementTypeOther/XCUIElementTypeOther/XCUIElementTypeOther/XCUIElementTypeTable/XCUIElementTypeCell/XCUIElementTypeStaticText[2]
```

And they are quite slow, since xpath is not natively supported by XCTest.

My implementation uses the same query syntax as xpath (only class names and indexes are supported though), but the search itself takes much less time in comparison to xpath, because it looks for direct child nodes only and uses single native method call (_childrenMatchingType_) for this purpose. Examples of valid class chain queries:

```
XCUIElementTypeWindow/XCUIElementTypeButton[3] - select the third child button of the first child window element (indexing starts from 1. zero index is invalid, negative indexes are allowed)

XCUIElementTypeWindow - select all the children windows

XCUIElementTypeWindow[2] - select the second child window in the hierarchy. Indexing starts at 1

XCUIElementTypeWindow/XCUIElementTypeAny[3] - select the third child (of any type) of the first child window

XCUIElementTypeWindow/*[3] - same as above

XCUIElementTypeWindow[2]/XCUIElementTypeAny - select all the children of the second child window

XCUIElementTypeWindow[2]/XCUIElementTypeAny[-2] - select the second last child of the second child window
```

I've also implemented nice descriptions for syntax errors.

### CODE REVIEW

**Comment by marekcirkos** [comment]
@mykola-mokhnach Can you email me or poke me on Facebook?

**Comment by mykola-mokhnach** [comment]
@marekcirkos Sorry, but I'm not registered on FB and not gonna do it ever.
I'd be happy to contact you via email, but don't know your address.

**Comment by marekcirkos** [comment]
marekc@fb.com

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+- (NSArray<XCUIElement *> *)fb_descendantsMatchingClassChain:(NSString *)classChainQuery shouldReturnAfterFirstMatch:(BOOL)shouldReturnAfterFirstMatch
+{
+  NSError *error;
+  FBClassChain parsedChain = [FBClassChainQueryParser parseQuery:classChainQuery error:&error];
+  if (nil == parsedChain) {
+    @throw [NSException exceptionWithName:FBClassChainQueryParseException reason:error.localizedDescription userInfo:error.userInfo];
+    return nil;
+  }
+  NSMutableArray *unmatchedSnapshots = [NSMutableArray arrayWithArray:[self fb_lookupChain:parsedChain root:self.fb_lastSnapshot]];
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: or `[self fb_lookupChain:parsedChain root:self.fb_lastSnapshot].mutableCopy`

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  NSMutableArray *result = [NSMutableArray array];
+  NSMutableArray *candidateElements = [NSMutableArray arrayWithArray:elementsQuery.allElementsBoundByIndex];
+  if ([parsedChain lastObject].position < 0 && candidateElements.count > 1) {
+    candidateElements = [[[candidateElements reverseObjectEnumerator] allObjects] mutableCopy];
+  }
+  NSUInteger candidateElementIdx = 0;
+  while (candidateElementIdx < candidateElements.count) {
+    XCUIElement *candidateElement = [candidateElements objectAtIndex:candidateElementIdx];
+    NSUInteger snapshotIdx = 0;
+    BOOL isMatchFound = NO;
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: This is always false. Doesn't seem like you need it

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        [unmatchedSnapshots removeObjectAtIndex:snapshotIdx];
+        break;
+      }
+      ++snapshotIdx;
+    }
+    if (0 == unmatchedSnapshots.count) {
+      break;
+    }
+    if (isMatchFound) {
+      [candidateElements removeObjectAtIndex:candidateElementIdx];
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: This never happens

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  for (NSUInteger level = 1; level < parsedChain.count; ++level) {
+    elementsQuery = [elementsQuery childrenMatchingType:[parsedChain objectAtIndex:level].type];
+  }
+  NSMutableArray *result = [NSMutableArray array];
+  NSMutableArray *candidateElements = [NSMutableArray arrayWithArray:elementsQuery.allElementsBoundByIndex];
+  if ([parsedChain lastObject].position < 0 && candidateElements.count > 1) {
+    candidateElements = [[[candidateElements reverseObjectEnumerator] allObjects] mutableCopy];
+  }
+  NSUInteger candidateElementIdx = 0;
+  while (candidateElementIdx < candidateElements.count) {
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: Fast enumeration should be better `for (element in candidateElements)`

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  }
+  return result.copy;
+}
+
+- (NSArray<XCElementSnapshot *> *)fb_lookupChain:(FBClassChain)query root:(XCElementSnapshot *)root
+{
+  FBClassChainElement *chainElement = [query firstObject];
+  NSArray *filteredChildren = root.children;
+  if (XCUIElementTypeAny != chainElement.type) {
+    filteredChildren = [filteredChildren filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"elementType == %@", @(chainElement.type)]];
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: Please use `FBStringify` instead of plain strings, https://github.com/facebook/WebDriverAgent/blob/2eefc210ab3a775cfd1eff1b47e1a81484b64b50/WebDriverAgentLib/Commands/FBFindElementCommands.m#L77

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ Instance constructor, which allows to set element type and position
+ 
+ @param type on of supoported element types declared in XCUIElementType enum
+ @param position element position relative to its sibling element. Numeration
+   starts with 1. Zero value means that all sibling element should be selected.
+   Negative value means that numeration starts from the last element, for example
+   -1 is the last child element and -2 is the second last element
+ @return FBClassChainElement instance
+ */
+- (id) initWithType:(XCUIElementType)type position:(int)position;
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.h]: Use `NSInteger` instead of `int`

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+NS_ASSUME_NONNULL_END
+
+
+@implementation Token
+
+- (id)init
+{
+  self = [super init];
+  if (self) {
+    self.asString = @"";
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: Do not call `self` in init. `_asString = @"";`

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+  }
+  return self;
+}
+
+- (id)initWithStringValue:(NSString *)stringValue
+{
+  self = [super init];
+  if (self) {
+    self.asString = stringValue;
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: Same here

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+  return self;
+}
+
++ (NSCharacterSet *)allowedCharacters
+{
+  // This method is expected to be overriden by subclasses
+  return [NSCharacterSet characterSetWithCharactersInString:@""];
+}
+
++ (unsigned int)maxLength
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: `NSUInteger`

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+  }
+  return self;
+}
+
+@end
+
+
+@implementation FBClassChainQueryParser
+
++ (void)tokenizationErrorWithIndex:(unsigned int)index originalQuery:(NSString *)originalQuery error:(NSError **)error
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: If you accept an error you should return a `BOOL` to indicate whether operation was successful.

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+{
+  NSString *description = [NSString stringWithFormat:@"Cannot parse class chain query '%@'. Unexpected character detected at position %d:\n'%@' <----", originalQuery, index + 1, [originalQuery substringToIndex:index + 1]];
+  *error = [[FBErrorBuilder.builder withDescription:description] build];
+}
+
++ (nullable NSArray<Token *> *)tokenizedQueryWithQuery:(NSString*)classChainQuery error:(NSError **)error
+{
+  unsigned int queryStringLength = (unsigned int)classChainQuery.length;
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: `NSUInteger`

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+ */
+
+#import "FBExceptionHandler.h"
+#import "FBClassChainQueryParser.h"
+#import "FBErrorBuilder.h"
+#import "FBElementTypeTransformer.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface Token : NSObject
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: You need `FB` prefix in all of your classes. Also I would be more precise and name it something like  `FBClassChainToken` ?

Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: This is an internal class, which is not used anywhere else, except of this particular implementation . Do we still need the prefix

Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: In Objc you need to always prefix class names.

**Comment by mykola-mokhnach** [comment]
> I am not a big fun of this approach. It is super sensitive to tree hierarchy changes and class name changes. Instead people should use explicit accessibility identifiers and the whole xpath stuff is not needed

Me too, but I still see many people are using terribly long xpath locators autogenerated by inspector, because either application developers are too lazy to assign identifiers or these people don't know Appium/WDA features well enough. Class chain locator usage could be kind of quick patch for lookup speed in such case.

**Comment by mykola-mokhnach** [comment]
@marekcirkos There are already unit tests for chain parser. Please check _FBClassChainTests.m_ in UnitTests folder.

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 1 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: done

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 2 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: Good catch. Fixed.

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 4 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: Now this is OK after isMatchFound behaviour is fixed

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 3 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: Fixed

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.h (hunk 6 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.h]: fixed

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m (hunk 7 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: fixed

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m (hunk 8 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: fixed

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m (hunk 9 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: fixed

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m (hunk 10 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: This is just a helper method to initialize NSError instance with extended description. Anyway, I've changed returned result type to BOOL.

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m (hunk 11 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: fixed

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 5 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: fixed. thx for the hint

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 4 above)
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: I am not talking about speed only. Also readability will be much better

Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: we need while loop here, because the _candidateElements_ array is changed while iteration is in progress. This won't work for _for .. in_ loop.

**Comment by mykola-mokhnach** [comment]
@marekcirkos Are there any new comments to this PR?

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.h [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+/*! Exception used to notify about invalid class chain query */
+extern NSString *const FBClassChainQueryParseException;
+
+@interface XCUIElement (FBClassChain)
+
+/**
+ Returns an array of descendants matching given class chain query. This type of queries is a subset of xpath queries, but works much faster because of the internal implementation features.
+ 
+ @param classChainQuery requested class chain query. This query is constructed like xpath, but can only include indexes and valid class names. Only search by direct children elements of the current element is supported. Examples of such requests:
+ XCUIElementTypeWindow/XCUIElementTypeButton[3] - select the third child button of the first child window element
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.h]: nit: I would keep `@param` short and keep juicy documentation above.

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.h [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+ Instance constructor, which allows to set element type and position
+ 
+ @param type on of supoported element types declared in XCUIElementType enum
+ @param position element position relative to its sibling element. Numeration
+   starts with 1. Zero value means that all sibling element should be selected.
+   Negative value means that numeration starts from the last element, for example
+   -1 is the last child element and -2 is the second last element
+ @return FBClassChainElement instance
+ */
+- (id) initWithType:(XCUIElementType)type position:(NSInteger)position;
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.h]: instancetype instead of id

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+}
+
++ (BOOL)compilationErrorWithQuery:(NSString *)originalQuery description:(NSString *)description error:(NSError **)error
+{
+  NSString *fullDescription = [NSString stringWithFormat:@"Cannot parse class chain query '%@'. %@", originalQuery, description];
+  *error = [[FBErrorBuilder.builder withDescription:fullDescription] build];
+  return YES;
+}
+
++ (nullable FBClassChain)compiledQueryWithTokenizedQuery:(NSArray<FBBaseClassChainToken *> *)tokenizedQuery originalQuery:(NSString *)originalQuery error:(NSError **)error
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: This is an `NSArray` not `FBClassChain`

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+#import "XCUIElement+FBUtilities.h"
+
+NSString *const FBClassChainQueryParseException = @"FBClassChainQueryParseException";
+
+@implementation XCUIElement (FBClassChain)
+
+- (NSArray<XCUIElement *> *)fb_descendantsMatchingClassChain:(NSString *)classChainQuery shouldReturnAfterFirstMatch:(BOOL)shouldReturnAfterFirstMatch
+{
+  NSError *error;
+  FBClassChain parsedChain = [FBClassChainQueryParser parseQuery:classChainQuery error:&error];
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: This `NSArray` not `FBClassChain`. `FBClassChain` is undefined, I can't find any definition.

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+#import "FBElementUtils.h"
+#import "FBMacros.h"
+#import "XCElementSnapshot.h"
+#import "XCUIElement+FBUtilities.h"
+
+NSString *const FBClassChainQueryParseException = @"FBClassChainQueryParseException";
+
+@implementation XCUIElement (FBClassChain)
+
+- (NSArray<XCUIElement *> *)fb_descendantsMatchingClassChain:(NSString *)classChainQuery shouldReturnAfterFirstMatch:(BOOL)shouldReturnAfterFirstMatch
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: This method is really hard to read. Can you break it to smaller methods that are more descriptive?

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 4 above)
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: Why don't you simply iterate `candidateElements`? This is effectively what is happening, but is really hard to follow when reading this code.

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+    if (isMatchFound) {
+      [candidateElements removeObjectAtIndex:candidateElementIdx];
+    } else {
+      ++candidateElementIdx;
+    }
+  }
+  return result.copy;
+}
+
+- (NSArray<XCElementSnapshot *> *)fb_lookupChain:(FBClassChain)query root:(XCElementSnapshot *)root
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: Please rename `fb_lookupChain` and break it to smaller peaces.

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.h [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+#import <XCTest/XCTest.h>
+
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FBClassChainElement : NSObject
+
+/*! Element's position */
+@property (readonly) NSInteger position;
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.h]: you should make all properties `nonatomic` unless you really, really want them `atomic`.

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.h (hunk 14 above)
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.h]: Kill space after `(id)`. I noticed it in few other places cross diff..

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m [hunk 20]
```diff
@@ -0,0 +1,373 @@
...
+  }
+  return self;
+}
+
+@end
+
+
+@implementation FBClassChainQueryParser
+
++ (BOOL)tokenizationErrorWithIndex:(NSUInteger)index originalQuery:(NSString *)originalQuery error:(NSError **)error
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: Why not return an error here, if it is a error constructor?

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m [hunk 21]
```diff
@@ -0,0 +1,373 @@
...
+  FBBaseClassChainToken *lastToken = [result lastObject];
+  if (!([lastToken isKindOfClass:FBClosingBracketToken.class] || [lastToken isKindOfClass:FBClassNameToken.class])) {
+    [self.class tokenizationErrorWithIndex:queryStringLength - 1 originalQuery:classChainQuery error:error];
+    return nil;
+  }
+  
+  return result.copy;
+}
+
++ (BOOL)compilationErrorWithQuery:(NSString *)originalQuery description:(NSString *)description error:(NSError **)error
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: Same here

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m [hunk 22]
```diff
@@ -0,0 +1,373 @@
...
+        }
+        @throw e;
+      }
+    } else if ([token isKindOfClass:FBNumberToken.class]) {
+      if (isPositionSet) {
+        NSString *description = [NSString stringWithFormat:@"Position value '%@' is expected to be set only once.", token.asString];
+        [self.class compilationErrorWithQuery:originalQuery description:description error:error];
+        return nil;
+      }
+      NSNumberFormatter *f = [[NSNumberFormatter alloc] init];
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: Formatters are expensive to create you should do this in constructor and reuse it.

File: WebDriverAgentTests/UnitTests/FBClassChainTests.m [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+  
+  FBClassChainElement *secondElement = [result objectAtIndex:1];
+  XCTAssertEqual(secondElement.type, XCUIElementTypeButton);
+  XCTAssertEqual(secondElement.position, -1);
+}
+
+- (void)testInvalidChains
+{
+  NSArray *invalidQueries = @[
+                              @"/XCUIElementTypeWindow"
```
Comment (marekcirkos) [inline WebDriverAgentTests/UnitTests/FBClassChainTests.m]: Fix spacing

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m (hunk 15 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: > +/*! Type alias for the product of class chain query parsing */
> +typedef NSArray<FBClassChainElement *> * FBClassChain;

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m (hunk 22 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: ok, thanks for the hint

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 4 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: > [candidateElements removeObjectAtIndex:candidateElementIdx];

because of that. I cannot change the structure being currently iterated, otherwise this will corrupt my iterator.

I want to remove matched items immediately to reduce iterations count

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 16 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: > +/*! Type alias for the product of class chain query parsing */
> +typedef NSArray<FBClassChainElement *> * FBClassChain;

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 17 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: sure will do

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.h (hunk 13 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.h]: fixed

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.h (hunk 19 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.h]: fixed

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.h (hunk 14 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.h]: fixed

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m (hunk 20 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: changed both method to return NSError* instead

File: WebDriverAgentLib/Utilities/FBClassChainQueryParser.m (hunk 21 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBClassChainQueryParser.m]: changed both method to return NSError* instead

File: WebDriverAgentTests/UnitTests/FBClassChainTests.m (hunk 23 above)
Comment (mykola-mokhnach) [inline WebDriverAgentTests/UnitTests/FBClassChainTests.m]: fixed

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m (hunk 18 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: done

File: WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+#import "XCUIElement+FBClassChain.h"
+#import "FBClassChainQueryParser.h"
+#import "FBElementUtils.h"
+#import "FBMacros.h"
+#import "XCElementSnapshot.h"
+#import "XCUIElement+FBUtilities.h"
+
+NSString *const FBClassChainQueryParseException = @"FBClassChainQueryParseException";
+
+@implementation XCUIElement (FBClassChain)
```
Comment (marekcirkos) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: Can add `fb_` prefix to all methods in that class? Let's land it then.

Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m]: added and squashed

**Comment by marekcirkos** [comment]
Travis is not happy, looks like `testSpecialXPathQuery` and https://travis-ci.org/facebook/WebDriverAgent/jobs/206118659 are broken.

**Comment by mykola-mokhnach** [comment]
Sorry, my bad. Both should be fixed now

**Comment by marekcirkos** [comment]
Finally! Thank you @mykola-mokhnach for your patience
